---
title: "Bladder cohort - R Notebook"
output:
  pdf_document: default
  html_document:
    df_print: paged
---
Adam Cankaya
<br>\
acankaya2017@fau.edu
<br>\
Based on tutorials https://www.costalab.org/wp-content/uploads/2020/11/R_class_D3.htm
and https://bioinformaticsworkbook.org/tutorials/wgcna.html#gsc.tab=0
<br>\

First install BiocManager, edgeR, and TCGAbiolinks
```{r}
ptm <- proc.time()  # start the timer

if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

BiocManager::install("edgeR")
BiocManager::install("TCGAbiolinks")
BiocManager::install("genefilter")
BiocManager::install("EDASeq")
# BiocManager::install("clusterProfiler")
# options(connectionObserver = NULL) # problem in loading databases using RSQLite
# BiocManager::install("org.Hs.eg.db")
install.packages(c("tidyverse", "magrittr"))
BiocManager::install("impute")
BiocManager::install("preprocessCore")
BiocManager::install("WGCNA")

chunk1_time = (proc.time() - ptm)[3] # stop the timer
```

<br>\
<br>\
TODO - rename steps to match steps from slides
Step 1 - Load packages, download data from TCGA, and prepare it for DEGList
```{r}
ptm <- proc.time()  # start the timer

library("TCGAbiolinks")
library("limma")
library("edgeR")
library("glmnet")
library("factoextra")
library("FactoMineR")
library("caret")
library("SummarizedExperiment")
library("gplots")
library("survival")
library("survminer")
library("RColorBrewer")
library("gProfileR")
library("genefilter")
# library("clusterProfiler")

setwd('C:/Adam/R/')  # make sure it already exists

ptm <- proc.time()  # start the timer

# Before we perform a GDC query let's look at the TCGA-BLCA data
# As of June 2024 we should see a case count of 412
TCGAbiolinks:::getProjectSummary("TCGA-BLCA")

# Download TCGA-BLCA data from GDC
# We want the complete RNA sequencing and raw gene count data 
# So we run a query of the Transcriptome Profiling category and RNA-Seq experimental type
# We use the STAR - Counts workflow type because it contains the raw gene counts we need
# We ignore other sample types besides tumor and normal
# The original paper by Wang uses the HTSeq-counts workflow, but this is a legacy version of
#  the new STAR - COUNTS workflow type
query_TCGA = GDCquery(
  project = "TCGA-BLCA",
  data.category = "Transcriptome Profiling",
  data.type="Gene Expression Quantification",
  experimental.strategy = "RNA-Seq",
  workflow.type = "STAR - Counts",
  sample.type = c("Primary Tumor", "Solid Tissue Normal"))

# Run the query and format it as a table
# The results are a table with 431 rows (because some patients have multiple cases each)
# There are 29 columns with meta data about each case such as sample_type (tumor vs normal)
lihc_res = getResults(query_TCGA)

# We can create a summary table shows there are 412 tumor and 19 normal (412+19=431)
summary(factor(lihc_res$sample_type))

# Go ahead and download all the data from GDC to our working directory
GDCdownload(query = query_TCGA)

# Now load the RNA-Seq data from the files into R workspace
tcga_data = GDCprepare(query_TCGA)

# This data object has 60660 rows and 431 columns
# This indicates there are 60660 different genes found throughout all the cases
# The object contains both clincal and expression data
dim(tcga_data)

# We can access the data in the object like this which verifies 412 tumor and 19 normal
table(tcga_data@colData$definition)

# Or see the gender data of 117 female and 314 male
table(tcga_data@colData$gender)

# let's look at the various names of the first 6 genes...
head(rowData(tcga_data))

# To preview the raw gene counts let's look at the expression levels of the first 
#  6 genes in the first 3 cases...
rownames = values(tcga_data)$gene_name[1:6]
firs6genes = head(assay(tcga_data)[,1:3])
rownames(firs6genes) = rownames
colnames(firs6genes) = c("Case 1", "Case 2", "Case 3")
firs6genes

chunk2_time = (proc.time() - ptm)[3] # stop the timer
```

<br>\
<br>\
Step 2 - Generate DGEList, filter low counts, and normalize data
```{r}
# Before we can perform DEG analysis we need to normalize the data
# Let's create a limma pipeline to do this...
# The pipeline function will take in three input parameters:
#   tcga_data - the data object we created in Step 1
#   condition_variable - the variable by which we will group patients (tumor vs normal)
#   reference_group - indicates which of the condition variable
#    values is the reference group (no tumors)
# The pipeline will return a list of three objects:
#   voom - the TMM normalized data returned by running voom
#   eBayes - the fitted model returned by running eBayes
#   topTable - a simple table which contains the top 1000 differentially expressed genes 
#    sorted by p.value
limma_pipeline = function(
  tcga_data,
  condition_variable,
  reference_group=NULL) {

    # Create a design matrix
    # The factor is the category classifier for the data (tumor vs normal)
    #  limma requires it to be a factor object
    design_factor = colData(tcga_data)[, condition_variable, drop=T] # definition
    group = factor(design_factor) # Solid Normal Tissue
    
    # otherwise just pick the first class as the reference class
    if (!is.null(reference_group)) {
      group = relevel(group, ref=reference_group) 
    }
  
    # make the design matrix
    design = model.matrix(~ group)
  
    # generate the DGEList object using the input...
    #  counts is the raw gene counts (numericla matrix - rows as genes, columns as cases)
    #  samples is the clinical data (data frame)
    #  genes is the annotation information (data frame - gene id and names)
    # the DGEList object returned is a transformed version of tcga_data
    dge = DGEList(counts=assay(tcga_data),
                   samples=colData(tcga_data),
                   genes=as.data.frame(rowData(tcga_data)))
  
    # filtering - by default genes with less than 10 counts per million reads are removed
    # after filtering we have 28087 genes remaining
    # no need to filter further by logfc or adjusted p-value because all
    #  entries already meet the cutoff criteria
    keep = filterByExpr(dge,design) # genes which meet are left after filtering
    dge = dge[keep,,keep.lib.sizes=FALSE] # filter the DGEList object, only keep the genes we want
    rm(keep) # remove this object from memory because we are done with it
    
    # TODO do we need rpkm() filtering?
  
    # Normalization (TMM followed by voom)
    # normalizing - minimize batch effects and variation with the TMM normalization
    # TMM - trimmed mean of M-values
    # use the voom method to convert the data to have a similar variance as arrays
    dge = calcNormFactors(dge)
    v = voom(dge, design, plot=TRUE)
  
    # Fit model to data given design
    #  fits a series of linear models, one to each probe
    #  then pass it to eBayes to rank the differential expression
    fit = lmFit(v, design)
    fit = eBayes(fit)
    
    # Save top genes
    topGenes = topTable(fit, coef=ncol(design), number=1000, sort.by="p")
  
    return(
      list(
        voomObj=v, # normalized data
        fit=fit, # fitted model and statistics
        topGenes=topGenes # the 1000 most differentially expressed genes
      )
    )
}

# Run the pipeline on the tcga_data from step 1 and normal tissue as the reference
#  "definition" is the column name for the tissue type (tumor vs normal)
#  "Solid Tissue Normal" is our baseline/control/reference class value
# The limma_res object returned is a list of 3 objects - voomObj, fit, topGenes
ptm <- proc.time()  # start the timer
limma_res = limma_pipeline(
  tcga_data=tcga_data,
  condition_variable="definition",
  reference_group="Solid Tissue Normal"
)
chunk3_time = (proc.time() - ptm)[3] # stop the timer
```

<br>\
<br>\
Step 3 - Visualize
```{r}
ptm <- proc.time()  # start the timer

# make a function to generate a scatter plot to show a separation of tumor vs normal points
plot_PCA = function(voomObj, condition_variable){
  # create a factor
  group = factor(voomObj$targets[, condition_variable])
  # perform a principal component analysis 
  pca = prcomp(t(voomObj$E))
  # Take PC1 and PC2 for the plot
  plot(pca$x[,1:2],col=group, pch=19)
  # include a legend for points
  legend("bottom", inset=.01, levels(group), pch=19, col=1:length(levels(group)))
  return(pca)
}

# call the plot function with the voom object and the defintion column
res_pca = plot_PCA(limma_res$voomObj, "definition")

# create a volcano plot
x = limma_res$topGenes$logFC
y = limma_res$topGenes$adj.P.Val
TCGAVisualize_volcano(
  x,
  y,
  xlab = "logFC",
  title = "Volcano plot of top 1000 genes",)

chunk4_time = (proc.time() - ptm)[3] # stop the timer
```

<br>\
<br>\
Step 4 - Classification model training, testing, and evaluation
```{r}
ptm <- proc.time()  # start the timer

# use the expression data that has been normalized
# Transpose and make it into a matrix object
input_mat = as.matrix(t(limma_res$voomObj$E))

# or try using the top 1000 genes
input_mat = as.matrix(t(limma_res$voomObj$E))

library(tidyverse)     # tidyverse will pull in ggplot2, readr, other useful libraries
library(magrittr)      # provides the %>% operator
library(WGCNA)
allowWGCNAThreads()          # allow up to 4 threads

# Choose a set of soft-thresholding powers
powers = c(c(1:10), seq(from = 12, to = 20, by = 2))

# Call the network topology analysis function
sft = pickSoftThreshold(
  input_mat,             # <= Input data
  #blockSize = 30,
  powerVector = powers,
  verbose = 5
  )

par(mfrow = c(1,2));
cex1 = 0.9;

plot(sft$fitIndices[, 1],
     -sign(sft$fitIndices[, 3]) * sft$fitIndices[, 2],
     xlab = "Soft Threshold (power)",
     ylab = "Scale Free Topology Model Fit, signed R^2",
     main = paste("Scale independence")
)
text(sft$fitIndices[, 1],
     -sign(sft$fitIndices[, 3]) * sft$fitIndices[, 2],
     labels = powers, cex = cex1, col = "red"
)
abline(h = 0.90, col = "red")
plot(sft$fitIndices[, 1],
     sft$fitIndices[, 5],
     xlab = "Soft Threshold (power)",
     ylab = "Mean Connectivity",
     type = "n",
     main = paste("Mean connectivity")
)
text(sft$fitIndices[, 1],
     sft$fitIndices[, 5],
     labels = powers,
     cex = cex1, col = "red")

picked_power = 5
temp_cor <- cor       
cor <- WGCNA::cor         # Force it to use WGCNA cor function (fix a namespace conflict issue)
netwk <- blockwiseModules(input_mat,                # <= input here

                          # == Adjacency Function ==
                          power = picked_power,                # <= power here
                          networkType = "signed",

                          # == Tree and Block Options ==
                          deepSplit = 2,
                          pamRespectsDendro = F,
                          # detectCutHeight = 0.75,
                          minModuleSize = 30,
                          maxBlockSize = 4000,

                          # == Module Adjustments ==
                          reassignThreshold = 0,
                          mergeCutHeight = 0.25,

                          # == TOM == Archive the run results in TOM file (saves time)
                          saveTOMs = T,
                          saveTOMFileBase = "ER",

                          # == Output Options
                          numericLabels = T,
                          verbose = 3)

cor <- temp_cor     # Return cor function to original namespace

# Convert labels to colors for plotting
mergedColors = labels2colors(netwk$colors)
# Plot the dendrogram and the module colors underneath
plotDendroAndColors(
  netwk$dendrograms[[1]],
  mergedColors[netwk$blockGenes[[1]]],
  "Module colors",
  dendroLabels = FALSE,
  hang = 0.03,
  addGuide = TRUE,
  guideHang = 0.05 )

# netwk$colors[netwk$blockGenes[[1]]]
# table(netwk$colors)

module_df <- data.frame(
  gene_id = names(netwk$colors),
  colors = labels2colors(netwk$colors)
)

module_df[1:5,]

write_delim(module_df,
            file = "gene_modules.txt",
            delim = "\t")

# Get Module Eigengenes per cluster
MEs0 <- moduleEigengenes(input_mat, mergedColors)$eigengenes

# Reorder modules so similar modules are next to each other
MEs0 <- orderMEs(MEs0)
module_order = names(MEs0) %>% gsub("ME","", .)

# Add case names
# TODO change this to cancer vs control
MEs0$case = row.names(MEs0)

# tidy & plot data
mME = MEs0 %>%
  pivot_longer(-case) %>%
  mutate(
    name = gsub("ME", "", name),
    name = factor(name, levels = module_order)
  )

mME %>% ggplot(., aes(x=case, y=name, fill=value)) +
  geom_tile() +
  theme_bw() +
  scale_fill_gradient2(
    low = "blue",
    high = "red",
    mid = "white",
    midpoint = 0,
    limit = c(-1,1)) +
  theme(axis.text.x = element_text(angle=90)) +
  labs(title = "Module-trait Relationships", y = "Modules", fill="corr")

# pick out a few modules of interest here
modules_of_interest = c("green", "turquoise", "tan")

# Pull out list of genes in that module
submod = module_df %>%
  subset(colors %in% modules_of_interest)

row.names(module_df) = module_df$gene_id

# # Get normalized expression for those genes
# expr_normalized[1:5,1:10]
# subexpr = expr_normalized[submod$gene_id,]
# 
# submod_df = data.frame(subexpr) %>%
#   mutate(
#     gene_id = row.names(.)
#   ) %>%
#   pivot_longer(-gene_id) %>%
#   mutate(
#     module = module_df[gene_id,]$colors
#   )
# 
# submod_df %>% ggplot(., aes(x=name, y=value, group=gene_id)) +
#   geom_line(aes(color = module),
#             alpha = 0.2) +
#   theme_bw() +
#   theme(
#     axis.text.x = element_text(angle = 90)
#   ) +
#   facet_grid(rows = vars(module)) +
#   labs(x = "treatment",
#        y = "normalized expression")

chunk5_time = (proc.time() - ptm)[3] # stop the timer
```

<br>\
<br>\
Step 5 - Hierarchical clustering
```{r}
# we are only considering the elastic net results to cluster genes together
# genes in green are original limma results
# genes in red are normal tissue from the elastic net results
# genes in black are tumor tissue from the elastic net results

ptm <- proc.time()  # start the timer

 # define the color palette for the plot
hmcol = colorRampPalette(rev(brewer.pal(9, "RdBu")))(256)

# perform complete linkage clustering
clust = function(x) hclust(x, method="complete")
# use the inverse of correlation as distance.
dist = function(x) as.dist((1-cor(t(x)))/2)

# Show green color for genes that also show up in DE analysis
colorLimmaGenes = ifelse(
  # Given a vector of boolean values
  (relevant_genes %in% limma_res$topGenes$ensembl_gene_id),
  "green", # if true, return green for that value
  "white" # if false, return white for that value
)

# generate the heatmap
gene_heatmap = heatmap.2(
  t(d_mat[,relevant_genes]),
  scale="row",          # scale the values for each gene (row)
  density.info="none",  # turns off density plot inside color legend
  trace="none",         # turns off trace lines inside the heat map
  col=hmcol,            # define the color map
  labRow=relevant_gene_names, # use gene names instead of ensembl annotation
  RowSideColors=colorLimmaGenes,
  labCol=FALSE,         # Not showing column labels
  ColSideColors=as.character(as.numeric(d_resp)), # Show colors for each response class
  dendrogram="both",    # Show dendrograms for both axis
  hclust = clust,       # Define hierarchical clustering method
  distfun = dist,       # Using correlation coefficient for distance function
  cexRow=.6,            # Resize row labels
  margins=c(1,5)        # Define margin spaces
)

chunk6_time = (proc.time() - ptm)[3] # stop the timer
```

<br>\
<br>\
Step 6 - Gene Ontology (GO) Enrichment Analysis (EA)
```{r}
# filter the relevant_gene_names by logFC value to find the up regulators
up_reg_genes = limma_res$topGenes[limma_res$topGenes$logFC < 1,]

# create the geneList as required by clusterProfiler
# save the gene name to first column and the logFC to second column
up_reg_genes_names = up_reg_genes[up_reg_genes$gene_name %in% relevant_gene_names,]$gene_name
up_reg_genes_logFC = up_reg_genes[up_reg_genes$gene_name %in% relevant_gene_names,]$logFC
geneList_up = up_reg_genes_logFC
names(geneList_up) = as.character(up_reg_genes_names)
geneList_up = sort(geneList_up, decreasing = TRUE)

ego3 <- gseGO(geneList     = geneList_up,
              OrgDb        = org.Hs.eg.db,
              ont          = "ALL",
              # minGSSize    = 100,
              # maxGSSize    = 500,
              # pvalueCutoff = 0.05,
              verbose      = TRUE,
              keyType      = "SYMBOL",
              # scoreType    = "pos"
              )

# and find the down regulators
down_reg_genes = limma_res$topGenes[limma_res$topGenes$logFC < -1,]
down_reg_genes = down_reg_genes[down_reg_genes$gene_name %in% relevant_gene_names,]$gene_name
```

Print out the timer values
```{r}
cat(paste0(
  "Chunk 1 finished in ", format(round(chunk1_time, 1), nsmall = 0), "s\n",
  "Chunk 2 finished in ", format(round(chunk2_time, 1), nsmall = 0), "s\n",
  "Chunk 3 finished in ", format(round(chunk3_time, 1), nsmall = 0), "s\n",
  "Chunk 4 finished in ", format(round(chunk4_time, 1), nsmall = 0), "s\n",
  "Chunk 5 finished in ", format(round(chunk5_time, 1), nsmall = 0), "s\n",
  "Chunk 6 finished in ", format(round(chunk6_time, 1), nsmall = 0), "s\n",
  "Total run time: ", format(round(chunk1_time+chunk2_time+
    chunk3_time+chunk4_time+chunk5_time+chunk6_time, 1), nsmall = 0), "s"))
```