---
title: "Bladder cohort - R Notebook"
output:
  pdf_document: default
  html_document:
    df_print: paged
---
Adam Cankaya
<br>\
acankaya2017@fau.edu
<br>\
Based on tutorials found at https://www.costalab.org/wp-content/uploads/2020/11/R_class_D3.htm
 and https://alexslemonade.github.io/refinebio-examples/04-advanced-topics/network-analysis_rnaseq_01_wgcna.html and https://github.com/hamidghaedi/RNA-seq-differential-expression
<br>\

First install packages
```{r}
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

# BiocManager::install("edgeR")
BiocManager::install("TCGAbiolinks")
# BiocManager::install("genefilter")
BiocManager::install("GO.db")
BiocManager::install("preprocessCore")
BiocManager::install("WGCNA")
BiocManager::install("DESeq2")

install.packages("ggforce")
```

<br>\
<br>\
Step 1 - Load packages, download data from TCGA, and prepare it for DESeq
```{r}
library("TCGAbiolinks")
library("limma")
# library("edgeR")
# library("glmnet")
# library("factoextra")
# library("FactoMineR")
# library("caret")
library("SummarizedExperiment")
# library("gplots")
# library("survival")
# library("survminer")
# library("RColorBrewer")
# library("gProfileR")
# library("genefilter")
# library("tidyverse")
library("magrittr")
library("WGCNA")
library("ggforce")
library("DESeq2")
library("doParallel")

setwd('C:/Adam/R/')  # make sure it already exists

# Before we perform a GDC query let's look at the TCGA-BLCA data
# As of June 2024 we should see a case count of 412
TCGAbiolinks:::getProjectSummary("TCGA-BLCA")

# TODO save data locally so we dont download it everytime

# Download TCGA-BLCA data from GDC
# We want the complete RNA sequencing and raw gene count data 
# So we run a query of the Transcriptome Profiling category and RNA-Seq experimental type
# We use the STAR - Counts workflow type because it contains the raw gene counts we need
# We ignore other sample types besides tumor and normal
# The original paper by Wang uses the HTSeq-counts workflow, but this is a legacy version of
#  the new STAR - COUNTS workflow type
query_TCGA = GDCquery(
  project = "TCGA-BLCA",
  data.category = "Transcriptome Profiling",
  data.type="Gene Expression Quantification",
  experimental.strategy = "RNA-Seq",
  workflow.type = "STAR - Counts",
  sample.type = c("Primary Tumor", "Solid Tissue Normal"))

# Run the query and format it as a table
# The results are a table with 431 rows (because some patients have multiple cases each)
# There are 29 columns with meta data about each case such as sample_type (tumor vs normal)
lihc_res = getResults(query_TCGA)

# We can create a summary table shows there are 412 tumor and 19 normal (412+19=431)
summary(factor(lihc_res$sample_type))

# Go ahead and download all the data from GDC to our working directory
GDCdownload(query = query_TCGA)

# Now load the RNA-Seq data from the files into R workspace
tcga_data = GDCprepare(query_TCGA)

# This data object has 60660 rows and 431 columns
# This indicates there are 60660 different genes found throughout all the cases
# The object contains both clinical and expression data
dim(tcga_data)

# We can access the data in the object like this which verifies 412 tumor and 19 normal
table(tcga_data@colData$definition)

# let's look at the first six rows (genes)
head(rowData(tcga_data))

# To preview the raw gene counts let's look at the expression levels of the first 
#  6 genes in the first 3 cases...
rownames = values(tcga_data)$gene_name[1:6]
firs6genes = head(assay(tcga_data)[,1:3])
rownames(firs6genes) = rownames
colnames(firs6genes) = c("Case 1", "Case 2", "Case 3")
firs6genes
```

<br>\
<br>\
Step 2 - Use DESeq to normalize data
```{r}
# exp matrix
rna <- as.data.frame(SummarizedExperiment::assay(tcga_data))
# clinical data
clinical <- data.frame(tcga_data@colData)

# replace spaces with "_" in levels of definition column
clinical$definition <-  gsub(" ", "_", clinical$definition)

# making the definition column as factor
clinical$definition <- as.factor(clinical$definition)
# relevling factor to ensure tumors would be compared to normal tissue.
levels(clinical$definition)
#
clinical$definition <- relevel(clinical$definition, ref = "Solid_Tissue_Normal")

# Making DESeqDataSet object which stores all experiment data
dds <- DESeqDataSetFromMatrix(countData = rna,
                              colData = clinical,
                              design = ~ definition)

dds_norm <- vst(dds)
```

<br>\
<br>\
Step 3 - Run WGCNA
```{r}
# Take our normalzied RNA-Seq data from DESeq
  normalized_counts <- assay(dds_norm) %>%
    t() # Transpose this data
  
# Calculate adjacensy threshold parameter value
  sft <- pickSoftThreshold(normalized_counts,
    dataIsExpr = TRUE,
    corFnc = cor,
    networkType = "signed"
  )
  
  # Calculate the signed R^2 (measure of model fit)
  sft_df <- data.frame(sft$fitIndices) %>%
    dplyr::mutate(model_fit = -sign(slope) * SFT.R.sq)
  
  # plot model fitting by the power soft threshold
  ggplot(sft_df, aes(x = Power, y = model_fit, label = Power)) +
  # Plot the points
  geom_point() +
  # We'll put the Power labels slightly above the data points
  geom_text(nudge_y = 0.1) +
  # We will plot what WGCNA recommends as an R^2 cutoff
  geom_hline(yintercept = 0.80, col = "red") +
  # Just in case our values are low, we want to make sure we can still see the 0.80 level
  ylim(c(min(sft_df$model_fit), 1.05)) +
  # We can add more sensible labels for our axis
  xlab("Soft Threshold (power)") +
  ylab("Scale Free Topology Model Fit, signed R^2") +
  ggtitle("Scale independence") +
  # This adds some nicer aesthetics to our plot
  theme_classic()
  
  # *** the plot shows a positive inflection around a soft threshold power of 12 so we use that ***

  # run WGCNA to find gene co-expression modules
  
  # use saved network if available
  if (file.exists("2025_05_18_wgcna_results.RDS")) {
      bwnet <- readRDS("2025_05_18_wgcna_results.RDS")
  } else {
      bwnet <- blockwiseModules(normalized_counts,
        maxBlockSize = 5000, # What size chunks (how many genes) the calculations should be run in
        TOMType = "signed", # topological overlap matrix
        power = 12, # soft threshold for network construction
        numericLabels = TRUE, # Let's use numbers instead of colors for module labels
        randomSeed = 1234, # there's some randomness associated with this calculation so we should set a seed
      )
      # save result for loading later
      write_rds(bwnet, "2025_05_18_wgcna_results.RDS")
  }
  
  module_eigengenes <- bwnet$MEs

  # Print out a preview
  head(module_eigengenes)
  
  # make sure samples are still in the same order
  all.equal(clinical$barcode, rownames(module_eigengenes))
  
  # Create the design matrix from the definition variable (tumor vs normal)
  des_mat <- model.matrix(~ clinical$definition)
  
  # Run linear model on each module. 
  # Limma wants our tests to be per row, so we also need to transpose so the eigengenes are rows
    # lmFit() needs a transposed version of the matrix
    fit <- limma::lmFit(t(module_eigengenes), design = des_mat)
    
    # Apply empirical Bayes to smooth standard errors
    fit <- limma::eBayes(fit)
    
  # Apply multiple testing correction and obtain stats in a data frame
    stats_df <- limma::topTable(fit, number = ncol(module_eigengenes)) %>%
      tibble::rownames_to_column("module")

  # look at the most significant modules
    head(stats_df)
    
  # *** The most significant module is ME6 so let's plot it***
    module_6_df <- module_eigengenes %>%
      tibble::rownames_to_column("barcode") %>%
    # Here we are performing an inner join with a subset of metadata
    dplyr::inner_join(clinical %>%
      dplyr::select(barcode, definition),
      by = c("barcode" = "barcode")
    )
    
  # *** the plot shows the module genes have elevated expression in normal tissue compared to tumor tissue ***
  ggplot(
    module_6_df,
    aes(
      x = definition,
      y = ME6,
      color = definition
    )
  ) +
  # a boxplot with outlier points hidden (they will be in the sina plot)
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  # A sina plot to show all of the individual data points
  ggforce::geom_sina(maxwidth = 0.3) +
  theme_classic()
  
  # *** Now let's look at what genes are in module 6 ***
  gene_module_key <- tibble::enframe(bwnet$colors, name = "gene", value = "module") %>%
    # Let's add the `ME` part so its more clear what these numbers are and it matches elsewhere
    dplyr::mutate(module = paste0("ME", module))
  gene_module_key %>%
    dplyr::filter(module == "ME6")
```

<br>\
<br>\
Step 4 and beyond - 
         a) experiment with parameters such as blocksize, soft threshold, more filtering...look at original paper's values
         b) examine all significant modules 
         c) Create heatmaps and other plots
         d) add timer code and speed enhancements (saving results to hd, multi-threading, AI?, cloud?, doParallel?)
         e) AI enhancements (Posit?, ask AI to speed up my code?)
         f) look at correlation strength (topological overlap matrix - TOMs)
         g) other TODOs
```{r}
```