---
title: "Bladder cohort - R Notebook"
output:
  pdf_document: default
  html_document:
    df_print: paged
---
Adam Cankaya
<br>\
acankaya2017@fau.edu
<br>\
Based on tutorials found at https://www.costalab.org/wp-content/uploads/2020/11/R_class_D3.htm
 and https://alexslemonade.github.io/refinebio-examples/04-advanced-topics/network-analysis_rnaseq_01_wgcna.html and https://github.com/hamidghaedi/RNA-seq-differential-expression

<br>\
<br>\
Step 1 - Install & load packages, download data from TCGA, and prepare it for DESeq
```{r}
r = getOption("repos")
r["CRAN"] = "http://cran.us.r-project.org"
options(repos = r)

if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

BiocManager::install("TCGAbiolinks")
BiocManager::install("GO.db")
BiocManager::install("preprocessCore")
BiocManager::install("WGCNA")
BiocManager::install("DESeq2")
BiocManager::install("org.Hs.eg.db")

install.packages("ggforce")

library("TCGAbiolinks")
library("limma")
library("SummarizedExperiment")
library(gprofiler2)
library("tidyverse")
library("magrittr")
library("WGCNA")
library("ggforce")
library("DESeq2")
library("doParallel")
library("org.Hs.eg.db")
library("AnnotationDbi")

setwd('C:/Adam/R/')  # make sure it already exists

if (file.exists("bladder_wgcna_workspace.RData")) {
  load("bladder_wgcna_workspace.RData") # load the saved workspace
}

# Before we perform a GDC query let's look at the TCGA-BLCA data
# As of June 2024 we should see a case count of 412
TCGAbiolinks:::getProjectSummary("TCGA-BLCA")

# Download TCGA-BLCA data from GDC
# We want the complete RNA sequencing and raw gene count data 
# So we run a query of the Transcriptome Profiling category and RNA-Seq experimental type
# We use the STAR - Counts workflow type because it contains the raw gene counts we need
# We ignore other sample types besides tumor and normal
# The original paper by Wang uses the HTSeq-counts workflow, but this is a legacy version of
#  the new STAR - COUNTS workflow type
query_TCGA = GDCquery(
  project = "TCGA-BLCA",
  data.category = "Transcriptome Profiling",
  data.type="Gene Expression Quantification",
  experimental.strategy = "RNA-Seq",
  workflow.type = "STAR - Counts",
  sample.type = c("Primary Tumor", "Solid Tissue Normal"))

# Run the query and format it as a table
# The results are a table with 431 rows (because some patients have multiple cases each)
# There are 29 columns with meta data about each case such as sample_type (tumor vs normal)
if (!exists("lihc_res") || is.null(lihc_res) || !length(lihc_res) || is_empty(lihc_res)) {
  lihc_res = getResults(query_TCGA)
  
  # We can create a summary table shows there are 412 tumor and 19 normal (412+19=431)
  summary(factor(lihc_res$sample_type))
  
  # Go ahead and download all the data from GDC to our working directory
  GDCdownload(query = query_TCGA)
  
  # Now load the RNA-Seq data from the files into R workspace which we will save to file later
  tcga_data = GDCprepare(query_TCGA)
}

# This data object has 60660 rows and 431 columns
# This indicates there are 60660 different genes found throughout all the cases
# The object contains both clinical and expression data
dim(tcga_data)

# We can access the data in the object like this which verifies 412 tumor and 19 normal
table(tcga_data@colData$definition)

# let's look at the first six rows (genes)
head(rowData(tcga_data))

# To preview the raw gene counts let's look at the expression levels of the first 
#  6 genes in the first 3 cases...
rownames = values(tcga_data)$gene_name[1:6]
firs6genes = head(assay(tcga_data)[,1:3])
rownames(firs6genes) = rownames
colnames(firs6genes) = c("Case 1", "Case 2", "Case 3")
firs6genes
```

<br>\
<br>\
Step 2 - Use DESeq to normalize data
```{r}
# rna expression data
rna <- as.data.frame(SummarizedExperiment::assay(tcga_data))
# clinical data
clinical <- data.frame(tcga_data@colData)

# replace spaces with "_" in levels of definition column
clinical$definition <-  gsub(" ", "_", clinical$definition)

# make the definition column a factor
clinical$definition <- as.factor(clinical$definition)
# relevel to ensure tumors are compared to normal tissue.
levels(clinical$definition)
clinical$definition <- relevel(clinical$definition, ref = "Solid_Tissue_Normal")

# Make DESeqDataSet object
dds <- DESeqDataSetFromMatrix(countData = rna,
                              colData = clinical,
                              design = ~ definition)
# Normalize data
dds_norm <- vst(dds)
```

<br>\
<br>\
Step 3 - Run WGCNA
```{r}
# generates plots for a given gene module
plot_module <- function(module_name) {
  module_df <- module_eigengenes %>%
    tibble::rownames_to_column("barcode") %>%
  # Here we are performing an inner join with a subset of metadata
  dplyr::inner_join(clinical %>%
    dplyr::select(barcode, definition),
    by = c("barcode" = "barcode")
  )
  
  print(
    ggplot(
      module_df,
      aes(
        x = definition,
        y = module_name,
        color = definition
      )
    ) +
    # a boxplot with outlier points hidden (they will be in the sina plot)
    geom_boxplot(width = 0.2, outlier.shape = NA) +
    # A sina plot to show all of the individual data points
    ggforce::geom_sina(maxwidth = 0.3) +
    theme_classic()
  )
  
  # *** Now let's look at what genes are in the module ***
  gene_module_key <- tibble::enframe(bwnet$colors, name = "gene", value = "module") %>%
    # Let's add the `ME` part so its more clear what these numbers are and it matches elsewhere
    dplyr::mutate(module = paste0("ME", module))
  gene_module_key %>%
    dplyr::filter(module == module_name)
  
  # *** For easy reading let's convert from ENSEMBLE to gene symbols ***
  positive_module_genes_ensembl <- gene_module_key[gene_module_key$module == module_name,]
  # convert to numeric to trim off the decimal place which represents the ensembl version
  positive_module_genes_ensembl_numeric <- sub("\\..*", "", positive_module_genes_ensembl$gene)
  # convert from ensembl to regular gene symbols and names
  positive_module_genes <- select(org.Hs.eg.db, keys = positive_module_genes_ensembl_numeric, 
                        keytype = 'ENSEMBL', columns = c('SYMBOL', "GENENAME"))
  # remove NA's
  positive_module_genes <- na.omit(positive_module_genes)
  print(head(positive_module_genes))
  
  # save as string for sending to AI
  positive_module_genes_string <- paste(positive_module_genes$SYMBOL, collapse=' ')
}


# Take normalized RNA-Seq data from DESeq
normalized_counts <- assay(dds_norm) %>%
  t() # Transpose this data

# Calculate adjacency threshold parameter value
if (!exists("sft") || is.null(sft) || !length(sft) || is_empty(sft)) {
  sft <- pickSoftThreshold(normalized_counts,
    dataIsExpr = TRUE,
    corFnc = cor,
    networkType = "signed"
  )
}
  
# Calculate the signed R^2 (measure of model fit)
sft_df <- data.frame(sft$fitIndices) %>%
  dplyr::mutate(model_fit = -sign(slope) * SFT.R.sq)
  
# plot model fitting by the power soft threshold
ggplot(sft_df, aes(x = Power, y = model_fit, label = Power)) +
# Plot the points
geom_point() +
# Put the Power labels  above the data points
geom_text(nudge_y = 0.1) +
# Plot what WGCNA recommends as an R^2 cutoff
geom_hline(yintercept = 0.80, col = "red") +
# Make sure we can still see low levels
ylim(c(min(sft_df$model_fit), 1.05)) +
# Adjust axis
xlab("Soft Threshold (power)") +
ylab("Scale Free Topology Model Fit, signed R^2") +
ggtitle("Scale independence") +
# Add some aesthetics
theme_classic()
  
# *** the plot shows a positive inflection around a soft threshold power of 12 so we use that ***

# run WGCNA to find gene co-expression modules

# use saved network if available
if (file.exists("2025_05_18_wgcna_results.RDS")) {
    bwnet <- readRDS("2025_05_18_wgcna_results.RDS")
} else {
    bwnet <- blockwiseModules(normalized_counts,
      maxBlockSize = 5000, # What size chunks (how many genes) the calculations should be run in
      TOMType = "signed", # topological overlap matrix
      power = 12, # soft threshold for network construction
      numericLabels = TRUE, # Let's use numbers instead of colors for module labels
      randomSeed = 1234, # there's some randomness associated with this calculation so we should set a seed
    )
    # save result for loading later
    readr::write_rds(bwnet, "2025_05_18_wgcna_results.RDS")
}

module_eigengenes <- bwnet$MEs
head(module_eigengenes)

# make sure samples are in same order
all.equal(clinical$barcode, rownames(module_eigengenes))

# Create the design matrix from the definition variable (tumor vs normal)
des_mat <- model.matrix(~ clinical$definition)

# Run linear model on each module. 
# limma wants our tests to be per row and lmFit() needs a transposed version of matrix
fit <- limma::lmFit(t(module_eigengenes), design = des_mat)
    
# Apply empirical Bayes to smooth standard errors
fit <- limma::eBayes(fit)
    
# Apply multiple testing correction and obtain stats in a data frame
stats_df <- limma::topTable(fit, number = ncol(module_eigengenes)) %>%
  tibble::rownames_to_column("module")

# look at the most significant modules
head(stats_df)
    
# *** The two most significant modules with negative log fc are ME6 & M10 (more associated with normal tissue)***
plot_module("ME6")
plot_module("ME10")

# *** The two most significant modules with positive log fc are ME7 & M54 (more associated with tumors)***
plot_module("ME7")
plot_module("ME54")
```
<br>\
<br>\
Step 4 and beyond - 
         a) experiment with parameters such as blocksize, soft threshold, more filtering...look at original paper's values
         b) examine all significant modules 
         c) Create heatmaps and other plots
         d) add timer code and speed enhancements (saving results to hd, multi-threading, AI?, cloud?, doParallel?)
         e) AI enhancements (Posit?, ask AI to speed up my code?)
         f) look at correlation strength (topological overlap matrix - TOMs)
         g) library(doParallel) registerDoParallel(cores=4)
         g) other TODOs
```{r}
  # Save the workspace for future loading
  save.image(file = "bladder_wgcna_workspace.RData")
```